### PR TITLE
Resolve database issues

### LINKED ISSUE #942: TC Tab build remains within the local storage of the database slot it was created in
Example, I have two databases:
image

There is no TC Tab build with the "frzyc" database:
image

For the sake of the example, there is a funny looking Raiden TC build within the "Stain" database:
image

I delete the "Stain" database:
image

Copy the "frzyc" database into slot2:
image

Now the "frzyc" database in slot 2 has the funny TC Tab build:
image

~~Discovered this because it seems that when a character has a TC Tab build present, changing talent dropdown settings causes the talent page to freeze.~~ #936

### LINKED ISSUE #955: One-time crash when uploading different DBs
1. Upload step 1 JSON (key thing here is that Aloy has some generated builds saved in the json)
2. Upload step 2 JSON over top of the same DB
3. Go to Aloy optimize tab
4. Crash
5. Reload, and crash is no longer there
[step1_2023-02-06_06-20-26.txt](https://github.com/frzyc/genshin-optimizer/files/10660276/step1_2023-02-06_06-20-26.txt)
[step2_2023-02-05_20-28-00.txt](https://github.com/frzyc/genshin-optimizer/files/10660278/step2_2023-02-05_20-28-00.txt)

### LINKED ISSUE #957: GO showing a confusing error when importing a bad Inventory Kamera scan
This is a Inventory Kamera error first and foremost. But the error handling could be improved by GO.

The attached database has heaps of 1* weapons with the "id" of 0.
This is causing a few issues;

One - it's doubling up equipped status on many characters:
``{"key":"FavoniusSword","level":90,"ascension":6,"refinement":1,"location":"Bennett","lock":true,"id":49},``
``{"key":"DullBlade","level":1,"ascension":0,"refinement":1,"location":"bennett","lock":false,"id":0},``

``{"key":"KagotsurubeIsshin","level":70,"ascension":4,"refinement":1,"location":"Jean","lock":true,"id":56},``
``{"key":"DullBlade","level":1,"ascension":0,"refinement":1,"location":"jean","lock":false,"id":0},``

Two - if the character actually has a 1* weapon equipped, for example Aloy only appears once in the json file, with the Hunter's Bow:

``{"key":"HuntersBow","level":1,"ascension":0,"refinement":1,"location":"aloy","lock":false,"id":0},``

That weapon selection is not being recognized so GO is assigning the default character archetype weapon (which happens to be Hunter's Bow)
However if you then re-import the same scan over the top - the import process complains saying it's removed an invalid Hunter's Bow:
image

Lastly - and this is really more for Inventory Kamera - the scanned account doesn't actually have heaps of 1* weapons:
image

[schnappi-genshinData_GOOD_2023_02_07_00_33.txt](https://github.com/frzyc/genshin-optimizer/files/10672518/schnappi-genshinData_GOOD_2023_02_07_00_33.txt)

### LINKED ISSUE #959: Make GO ids persistent
Currently, GOOD uses a non-id based system for art/char/weapons, this is usually good enough for dup/upgrade detection when merging databases. However, as GO have more and more relational data that depends on ids, this is no longer a sustainable model.

The proposal here is to make it that we maintain as much of the GO ids as possible. There are a few aspect to this:

1. There is a id migration mechanism that are not sustainable to maintain.
2. In the case of a GO onto GO import, the incoming ids should be respected, because its relational data will overwrite the current GO.
3. When relational data is broken, it is the `validation` function's responsibility to ensure relational data is rational.

### PR DESCRIPTION
**Pull request by frzyc**
* Mainly implement and resolve #959,
* Should also fix #957, fix #955 and fix #942

### CODE REVIEW

File: apps/frontend/src/app/Database/DataManagers/ArtifactData.ts [hunk 1]
```diff
@@ -1,292 +1,311 @@
...
+      let { duplicated, upgraded } = result.ignoreDups ? { duplicated: [], upgraded: [] } : this.findDups(art)
+      // Don't reuse dups/upgrades
+      duplicated = duplicated.filter(a => idsToRemove.has(a.id))
+      upgraded = upgraded.filter(a => idsToRemove.has(a.id))
+      if (duplicated[0] || upgraded[0]) {
+        // Favor upgrades with the same location, else use 1st dupe
+        const match = (hasEquipment && art.location && upgraded[0]?.location === art.location) ? upgraded[0] : (duplicated[0] || upgraded[0])
+        idsToRemove.delete(match.id)
+        if (duplicated[0]) result.artifacts.unchanged.push(art)
+        else if (upgraded[0]) result.artifacts.upgraded.push(art)
```
Comment (nguyentvan7) [inline apps/frontend/src/app/Database/DataManagers/ArtifactData.ts:105]: Based on if-condition on line 100, one of lines 104 or 105 is guaranteed to run, so we can simplify this

**Review by nguyentvan7 — CHANGES_REQUESTED** [review]
#405 is not resolved, since there is no additional text explaining why the items are marked as invalid

**Comment by frzyc** [comment]
> #405 is not resolved, since there is no additional text explaining why the items are marked as invalid

I'm not adding like dozens of text for somewhat self-explainable invalidations, and then getting translators to translate them. Not worth our time imo.

**Comment by nguyentvan7** [comment]
Ok then I'll remove #405 and close it separately

**Review by nguyentvan7 — APPROVED** [review]

**Comment by nguyentvan7** [comment]
Oh I see why 405 was there. Issue title is misleading since it's 2 issues in 1. I'll still close separately

**Pull request merged**